File: segundoPeriodo/AEDS-2/estudos/TADsFlexiveis/matriz/java/Main.java
import java.util.Scanner;

class Celula {
    public int elemento;
    public Celula sup, inf, prox, ant;

    public Celula() {
        this.elemento = 0;
        this.sup = null;
        this.inf = null;
        this.prox = null;
        this.ant = null;
    }
    public Celula(int x) {
        this.elemento = x;
        this.sup = null;
        this.inf = null;
        this.prox = null;
        this.ant = null;
    }
}

class Matriz {
    private Celula inicio;
    private int linha, coluna;

    public Matriz() {
        this(3, 3);
    }
    public Matriz(int l, int c) {
        this.linha = l;
        this.coluna = c;
        this.inicio = new Celula();
        Celula tmp = inicio;
        for(int i = 1; i < c; i++, tmp = tmp.prox) {
            tmp.prox = new Celula();
            tmp.prox.ant = tmp;
        }
        Celula col = inicio, linha = inicio;
        for(int i = 1; i < l; i++) {
            linha.inf = new Celula();
            linha.inf.sup = linha;
            linha = linha.inf;
            col = linha;
            for(int j = 1; j < c; j++, col = col.prox) {
                col.prox = new Celula();
                col.prox.ant = col;
                col.prox.sup = col.sup.prox;
                col.prox.sup.inf = col.prox;
            }
        }
    }

    public void inserir(Scanner sc) {
        Celula c = inicio, l = inicio;
        for(int i = 0; i < linha; i++) {
            for(int j = 0; j < coluna; j++) {
                c.elemento = sc.nextInt();
                c = c.prox;
            }
            l = l.inf;
            c = l;
        }
    }
}

public class Main {
    public static void main(String args[]) {
        Matriz matriz = new Matriz();
        Scanner sc = new Scanner(System.in);
        matriz.inserir(sc);
    }
}

File: segundoPeriodo/AEDS-2/estudos/TADsFlexiveis/matriz/c/main.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Celula {
    int elemento;
    struct Celula *inf, *sup, *esq, *dir;
} Celula;

Celula *criarCelula(int x) {
    Celula *nova = (Celula*)malloc(sizeof(Celula));
    nova->elemento = x;
    nova->inf = nova->sup = nova->esq = nova->dir = NULL;
    return nova;
}

typedef struct Matriz {
    int linhas, colunas;
    Celula *inicio;
} Matriz;

Matriz *criarMatriz(int l, int c) {
    Matriz *nova = (Matriz*)malloc(sizeof(Matriz));
    nova->linhas = l;
    nova->colunas = c;
    nova->inicio = criarCelula(0);
    Celula *tmp = nova->inicio;
    for(int i = 1; i < c; i++, tmp = tmp->dir) {
        tmp->dir = criarCelula(0);
        tmp->dir->esq = tmp;
    }
    Celula *lin = nova->inicio, *col = nova->inicio;
    for(int i = 1; i < l; i++) {
        lin->inf = criarCelula(0);
        lin->inf->sup = lin;
        lin = lin->inf;
        col = lin;
        for(int j = 1; j < c; j++) {
            col->dir = criarCelula(0);
            col->dir->esq = col;
            col = col->dir;
            col->sup = col->esq->sup->dir;
            col->sup->inf = col;
        }
    }
    return nova;
}

void inserir(Matriz *matriz) {
    Celula *l = matriz->inicio, *c = matriz->inicio;
    for(int i = 0; i < matriz->linhas; i++) {
        for(int j = 0; j < matriz->colunas; j++, c = c->dir) {
            scanf("%d", &c->elemento);
        }
        l = l->inf;
        c = l;
    }
}

void mostrar(Matriz *matriz) {
    Celula *l = matriz->inicio, *c;
    for(int i = 0; i < matriz->linhas; i++) {
        c = l;
        for(int j = 0; j < matriz->colunas; j++, c = c->dir) {
            printf("%d ", c->elemento);
        }
        printf("\n");
        l = l->inf;
    }
    c = l = NULL;
}

void removerUltimaColuna(Matriz *matriz) {
    Celula *l = matriz->inicio, *c;
    if(matriz->colunas == 1) {
        while(l != NULL) {
            c = l;
            l = l->inf;
            free(c);
        }
    }
    else {
        while(l != NULL) {
            for(c = l; c->dir != NULL; c = c->dir);
            c->esq->dir = NULL;
            c->esq = NULL;
            free(c);
            l = l->inf; 
        }
    }
    matriz->colunas--;
}

void freeMatriz(Matriz *matriz) {
    Celula *l = matriz->inicio, *c;
    while(l != NULL) {
        c = l;
        l = l->inf;
        while(c != NULL) {
            Celula *tmp = c->dir;
            free(c);
            c = tmp;
        }
    }
    free(matriz);
}

int main() {
    Matriz *matriz = criarMatriz(3, 3);
    inserir(matriz);
    mostrar(matriz);
    removerUltimaColuna(matriz);
    mostrar(matriz);
    freeMatriz(matriz);
    return 0;
}